### DISCUSSION TITLE
Sending commands from shell script to Pinnacle

### DISCUSSION THREAD

**Discussion by justinesmithies** [pinnacle-comp/pinnacle#217] [Q&A] [answered]
Is it possible to send commands ran from a bash script as you can using say `swaymsg` with Pinnacle ?
Thanks in advance and thank you for what looks to be a really fantastic tiling wm. I've been reading through the api and I can't wait to try it on my Void system.

**Comment by Ottatop** [comment, marked as answer]
> Is it possible to send commands ran from a bash script as you can using say swaymsg with Pinnacle ?

Yes, but you'll need to use the Lua API and either
1. Pipe Lua code into `lua`:
    ```sh
    echo "<your Lua here>" | lua
    ```
    or
2. Run Lua files from bash:
    ```sh
    lua <your Lua file>
    ```

If you're not limited to bash you can also run executable Lua files directly.

For the specifics, you can run the `run` function (not `setup`, this will block):
```lua
require("pinnacle").run(function(Pinnacle)
    -- Anything you can run in a normal Lua config you can use here
end)
```

Just note the following caveats:
- Setting a keybind will overwrite any current keybinds set, and because `run` returns you'll lose those keybinds. This may be changed in the future to allow for use cases like setting a keybind outside the config.
- You can't connect to signals, again because the script will return.
- You won't be able to get output from spawned processes for the same reason.

If you need to print output from the script in json you can install `cjson` from Luarocks:
```sh
luarocks install --local lua-cjson
```
and print tables as json with
```lua
print(require("cjson").encode(<your table here>))
```
